File: src/csharp/Foundgine.Core/Semantic/Authorization/SemanticAuthorizationContractValidator.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.IR;

namespace Foundgine.Core.Semantic.Authorization;

/// <summary>
/// Validates that an operation is rooted entirely in the trusted semantic
/// contract before authorization policy evaluation begins.
/// </summary>
internal static class SemanticAuthorizationContractValidator
{
    public static void Validate(SemanticContractSnapshot contract, SemanticOperation operation)
    {
        ArgumentNullException.ThrowIfNull(contract);
        ArgumentNullException.ThrowIfNull(operation);

        var visited = new HashSet<int>();
        ValidateNode(contract, operation.Root, visited, isRoot: true, parent: null);
    }

    private static void ValidateNode(
        SemanticContractSnapshot contract,
        SemanticReadNode node,
        ISet<int> visited,
        bool isRoot,
        SemanticEntity? parent)
    {
        if (!visited.Add(node.Id))
            throw new InvalidOperationException($"Semantic operation contains a cycle or duplicate node at {node.Id}.");

        if (!contract.TryGet(node.EntityId, out var entity))
            throw new InvalidOperationException(
                $"Semantic operation node {node.Id} references unknown entity '{node.EntityId}'.");

        if (isRoot && (node.ViaRelationship is not null || node.ViaConnection is not null))
            throw new InvalidOperationException($"Root semantic node {node.Id} cannot specify a parent edge.");

        if (!isRoot && node.ViaRelationship is null && node.ViaConnection is null)
            throw new InvalidOperationException(
                $"Non-root semantic node {node.Id} must specify the relationship or connection used to reach it.");

        if (node.ViaRelationship is { } relationshipId)
        {
            if (node.ViaConnection is not null)
                throw new InvalidOperationException(
                    $"Semantic node {node.Id} cannot specify both a relationship and a connection.");

            if (parent is not null)
            {
                var relationship = parent.Relationships.FirstOrDefault(x => x.Id == relationshipId)
                                   ?? throw new InvalidOperationException(
                                       $"Semantic operation node {node.Id} references relationship '{relationshipId}' not declared on '{parent.Name}'.");

                if (relationship.Target != node.EntityId)
                    throw new InvalidOperationException(
                        $"Semantic operation node {node.Id} targets '{node.EntityId}', but relationship '{relationship.Name}' targets '{relationship.Target}'.");
            }
        }

        foreach (var field in node.Fields.Concat(node.RequiredFields).Distinct())
        {
            if (field != entity.Identity.FieldId && entity.Fields.All(x => x.Id != field))
                throw new InvalidOperationException(
                    $"Semantic operation node {node.Id} selects unknown field '{field}' on '{entity.Name}'.");
        }

        foreach (var child in node.Children)
            ValidateNode(contract, child, visited, isRoot: false, parent: entity);
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/authorization/SemanticAuthorizationContractValidator.java
package com.foundgine.core.semantic.authorization;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.*;
import com.foundgine.core.semantic.ir.*;
import java.util.*;

/**
 * Validates canonical semantic IR against the trusted frozen contract before
 * policy evaluation.
 */
final class SemanticAuthorizationContractValidator {
	private SemanticAuthorizationContractValidator() {
	}

	static void validate(SemanticContractSnapshot contract, SemanticOperation operation) {
		Objects.requireNonNull(contract);
		Objects.requireNonNull(operation);
		validateNode(contract, operation.root(), new HashSet<>(), true, null);
	}

	private static void validateNode(SemanticContractSnapshot c, SemanticReadNode n, Set<Integer> visited, boolean root,
			SemanticEntity parent) {
		if (!visited.add(n.id()))
			throw new IllegalArgumentException(
					"Semantic operation contains a cycle or duplicate node at " + n.id() + ".");
		var entity = c.get(n.entityId());
		if (root && (n.viaRelationship() != null || n.viaConnection() != null))
			throw new IllegalArgumentException("Root semantic node " + n.id() + " cannot specify a parent edge.");
		if (!root && n.viaRelationship() == null && n.viaConnection() == null)
			throw new IllegalArgumentException("Non-root semantic node " + n.id()
					+ " must specify the relationship or connection used to reach it.");
		if (n.viaRelationship() != null) {
			if (n.viaConnection() != null)
				throw new IllegalArgumentException(
						"Semantic node " + n.id() + " cannot specify both a relationship and a connection.");
			if (parent != null) {
				var rel = parent.relationships().stream().filter(x -> x.id().equals(n.viaRelationship())).findFirst()
						.orElseThrow(() -> new IllegalArgumentException(
								"Semantic operation node " + n.id() + " references relationship '" + n.viaRelationship()
										+ "' not declared on '" + parent.name() + "'."));
				if (!rel.target().equals(n.entityId()))
					throw new IllegalStateException("Semantic operation node " + n.id() + " targets '" + n.entityId()
							+ "', but relationship '" + rel.name() + "' targets '" + rel.target() + "'.");
			}
		}
		var fields = new ArrayList<FieldId>();
		fields.addAll(n.fields());
		fields.addAll(n.requiredFields());
		for (var f : new LinkedHashSet<>(fields))
			if (!f.equals(entity.identity().fieldId()) && entity.fields().stream().noneMatch(x -> x.id().equals(f)))
				throw new IllegalArgumentException("Semantic operation node " + n.id() + " selects unknown field '" + f
						+ "' on '" + entity.name() + "'.");
		for (var child : n.children())
			validateNode(c, child, visited, false, entity);
	}
}
